File: VB/Scaffolding.ValidationErrors/Common/ViewModel/ReadOnlyLookUpCollectionViewModel.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Linq.Expressions
Imports DevExpress.Mvvm
Imports Scaffolding.ValidationErrors.Common.Utils
Imports Scaffolding.ValidationErrors.Common.DataModel

Namespace Scaffolding.ValidationErrors.Common.ViewModel
	Partial Public Class ReadOnlyLookUpCollectionViewModel(Of TMasterEntity As Class, TEntity As Class, TUnitOfWork As IUnitOfWork)
		Inherits ReadOnlyLookUpCollectionViewModelBase(Of TMasterEntity, TEntity, TUnitOfWork)
		Public Sub New(ByVal unitOfWorkFactory As IUnitOfWorkFactory(Of TUnitOfWork), ByVal getRepositoryFunc As Func(Of TUnitOfWork, IReadOnlyRepository(Of TEntity)), ByVal filterExpression As Expression(Of Func(Of TEntity, Boolean)))
			MyBase.New(unitOfWorkFactory, getRepositoryFunc, filterExpression)
		End Sub
	End Class
	Public MustInherit Class ReadOnlyLookUpCollectionViewModelBase(Of TMasterEntity As Class, TEntity As Class, TUnitOfWork As IUnitOfWork)
		Inherits ReadOnlyCollectionViewModel(Of TEntity, TUnitOfWork)

		Public Sub New(ByVal unitOfWorkFactory As IUnitOfWorkFactory(Of TUnitOfWork), ByVal getRepositoryFunc As Func(Of TUnitOfWork, IReadOnlyRepository(Of TEntity)), ByVal filterExpression As Expression(Of Func(Of TEntity, Boolean)))
			MyBase.New(unitOfWorkFactory, getRepositoryFunc, filterExpression)
		End Sub
	End Class
End Namespace

File: CS/Scaffolding.ValidationErrors/Common/ViewModel/ReadOnlyLookUpCollectionViewModel.cs
using System;
using System.Linq.Expressions;
using DevExpress.Mvvm;
using Scaffolding.ValidationErrors.Common.Utils;
using Scaffolding.ValidationErrors.Common.DataModel;

namespace Scaffolding.ValidationErrors.Common.ViewModel {
    public partial class ReadOnlyLookUpCollectionViewModel<TMasterEntity, TEntity, TUnitOfWork> : ReadOnlyLookUpCollectionViewModelBase<TMasterEntity, TEntity, TUnitOfWork>
        where TMasterEntity : class
        where TEntity : class
        where TUnitOfWork : IUnitOfWork {
        public ReadOnlyLookUpCollectionViewModel(IUnitOfWorkFactory<TUnitOfWork> unitOfWorkFactory, Func<TUnitOfWork, IReadOnlyRepository<TEntity>> getRepositoryFunc, Expression<Func<TEntity, bool>> filterExpression)
            : base(unitOfWorkFactory, getRepositoryFunc, filterExpression) {
        }
    }
    public abstract class ReadOnlyLookUpCollectionViewModelBase<TMasterEntity, TEntity, TUnitOfWork> : ReadOnlyCollectionViewModel<TEntity, TUnitOfWork>
        where TMasterEntity : class
        where TEntity : class
        where TUnitOfWork : IUnitOfWork {

        public ReadOnlyLookUpCollectionViewModelBase(IUnitOfWorkFactory<TUnitOfWork> unitOfWorkFactory, Func<TUnitOfWork, IReadOnlyRepository<TEntity>> getRepositoryFunc, Expression<Func<TEntity, bool>> filterExpression)
            : base(unitOfWorkFactory, getRepositoryFunc, filterExpression) {
        }
    }
}